lib/pair.ts

```diff
@@ -0,0 +1,266 @@
+  ): Promise<void>;
+
+  refund(amount: BigNumberish): Promise<void>;
+}
+
+export class ReadOnlyOptionPair implements ReadOptionPair {
+  protected option: Option;
+  protected obligation: Obligation;
+  protected optionLib: OptionLib;
+  protected collateral: Ierc20;
+  protected treasury: ITreasury;
+
+  constructor(
+    option: Option,
+    obligation: Obligation,
+    optionLib: OptionLib,
+    collateral: Ierc20,
+    treasury: ITreasury
+  ) {
+    this.option = option;
+    this.obligation = obligation;
+    this.optionLib = optionLib;
+    this.collateral = collateral;
+    this.treasury = treasury;
+  }
+
+  async getDetails(): Promise<{
+    expiryTime: number;
+    windowSize: number;
+    strikePrice: BigNumber;
+  }> {
+    const [expiryTime, windowSize, strikePrice] = await Promise.all([
+      this.option.expiryTime(),
+      this.option.windowSize(),
+      this.obligation.strikePrice()
+    ]);
+    return {
+      expiryTime: expiryTime.toNumber(),
+      windowSize: windowSize.toNumber(),
+      strikePrice
+    };
+  }
+
+  async getObligation(feed: BitcoinPriceFeed): Promise<EuropeanOption> {
+    return {
+      contract: this.obligation.address,
+      spotPrice: await feed.getSpotPrice(),
+      ...(await this.getDetails())
+    };
+  }
+
+  async getOption(feed: BitcoinPriceFeed): Promise<EuropeanOption> {
+    return {
+      contract: this.option.address,
+      spotPrice: await feed.getSpotPrice(),
+      ...(await this.getDetails())
+    };
+  }
+
+  async getBtcAddress(
+    account: string,
+    network: bitcoin.Network
+  ): Promise<string> {
+    const {btcHash, format} = await this.obligation.getBtcAddress(account);
+    return encodeBtcAddress(btcHash.substr(2), format, network)!;
+  }
+
+  // gets the locked collateral for a pair
+  async balanceOf(account: string): Promise<BigNumber> {
+    return this.treasury.balanceOf(this.obligation.address, account);
+  }
+}
+
+export class ReadWriteOptionPair extends ReadOnlyOptionPair
+  implements WriteOptionPair {
+  readonly account: string;
+  readonly confirmations?: number;
+
+  constructor(
+    option: Option,
+    obligation: Obligation,
+    optionLib: OptionLib,
+    collateral: Ierc20,
+    treasury: ITreasury,
+    account: string,
+    confirmations?: number
+  ) {
+    super(option, obligation, optionLib, collateral, treasury);
+    this.account = account;
+    this.confirmations = confirmations;
+  }
+
+  // back options with the default collateral
+  // this also adds liquidity to the uniswap pool
+  async write(
+    premium: BigNumberish,
+    amount: BigNumberish,
+    btcAddress: BtcAddress
+  ): Promise<void> {
+    await this.optionLib
+      .lockAndWrite(
+        this.option.address,
+        this.collateral.address,
+        this.collateral.address,
+        amount,
+        premium,
+        amount,
+        premium,
+        btcAddress.btcHash,
+        btcAddress.format
+      )
+      .then((tx) => tx.wait(this.confirmations));
```

Comment:
If we do this wouldn't it be a pain to show the user in the UI that the transaction has been submitted and is pending for confirmation? If so, I would rather suggest having something which able to register `on('confirmation', receipt => /* do some stuff */)` or so.